integration/kotlinx-io-okio/common/src/OkioAdapters.kt

```diff
@@ -0,0 +1,198 @@
+import kotlinx.io.unsafe.UnsafeBufferOperations
+import okio.ByteString.Companion.toByteString
+import kotlin.math.min
+
+private inline fun <T> withOkio2KxIOExceptionMapping(block: () -> T): T {
+    try {
+        return block()
+    } catch (eofe: okio.EOFException) {
+        throw kotlinx.io.EOFException(eofe.message)
+    } catch (ioe: okio.IOException) {
+        throw kotlinx.io.IOException(ioe.message, ioe)
+    }
+}
+
+private inline fun <T> withKxIO2OkioExceptionMapping(block: () -> T): T {
+    try {
+        return block()
+    } catch (eofe: kotlinx.io.EOFException) {
+        throw okio.EOFException(eofe.message)
+    } catch (ioe: kotlinx.io.IOException) {
+        throw okio.IOException(ioe.message, ioe)
+    }
+}
+
+/**
+ * Returns a [okio.Source] backed by this [kotlinx.io.RawSource].
+ *
+ * Closing one of these sources will also close another one.
+ */
+public fun RawSource.asOkioSource(): okio.Source = object : okio.Source {
+    private val buffer = Buffer() // TODO: optimization - reuse Source's buffer if possible
+
+    override fun close() = withKxIO2OkioExceptionMapping {
+        this@asOkioSource.close()
+    }
+
+    override fun read(sink: okio.Buffer, byteCount: Long): Long = withKxIO2OkioExceptionMapping {
+        val read = this@asOkioSource.readAtMostTo(buffer, byteCount)
+        if (read <= 0) return read
+
+        while (!buffer.exhausted()) {
+            @OptIn(UnsafeIoApi::class)
+            UnsafeBufferOperations.readFromHead(buffer) { data, from, to ->
+                val len = to - from
+                sink.write(data, from, len)
+                len
+            }
+        }
+        return read
+    }
+
+    override fun timeout(): okio.Timeout = okio.Timeout.NONE
+}
+
+/**
+ * Returns a [okio.Sink] backed by this [kotlinx.io.RawSink].
+ *
+ * Closing one of these sinks will also close another one.
+ */
+public fun RawSink.asOkioSink(): okio.Sink = object : okio.Sink {
+    private val buffer = Buffer() // TODO: optimization - reuse Sink's buffer if possible
+
+    override fun close() = withKxIO2OkioExceptionMapping {
+        this@asOkioSink.close()
+    }
+
+    override fun flush() = withKxIO2OkioExceptionMapping {
+        this@asOkioSink.flush()
+
+    }
+
+    override fun timeout(): okio.Timeout = okio.Timeout.NONE
+
+    override fun write(source: okio.Buffer, byteCount: Long) = withKxIO2OkioExceptionMapping {
+        require(source.size >= byteCount) {
+            "Buffer does not contain enough bytes to write. Requested $byteCount, actual size is ${source.size}"
+        }
+        var remaining = byteCount
+        while (remaining > 0) {
+            @OptIn(UnsafeIoApi::class)
+            UnsafeBufferOperations.writeToTail(buffer, 1) { data, from, to ->
+                val toWrite = min((to - from).toLong(), remaining).toInt()
+
+                val read = source.read(data, from, toWrite)
+                check(read == toWrite) { "Can't read $toWrite bytes from okio.Buffer" }
```

Comment:
This check is a bit aggressive as it relies on an implementation detail of `okio.Buffer` and is not written to the spec. I think all we care about here is looking for `-1`. `source.read` is within its rights (even on `Buffer`) to write less than `toWrite` into `data`. That will _never_ happen, of course, but I don't think we need to check that case, only instead that forward progress is being made.

This applies to the inverse adapter below, too.